bofire/surrogates/mixed_tanimoto_gp.py

```diff
@@ -0,0 +1,369 @@
+
+            covar_module = sum_kernel + prod_kernel
+
+        else:
+            sum_kernel = (
+                ScaleKernel(
+                    cont_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(ord_dims),
+                        active_dims=ord_dims,
+                    )
+                )
+                + ScaleKernel(
+                    mol_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(mol_dims),
+                        active_dims=mol_dims,
+                    )
+                )
+                + ScaleKernel(
+                    CategoricalKernel(
+                        batch_shape=aug_batch_shape,
+                        ard_num_dims=len(cat_dims),  # type: ignore
+                        active_dims=cat_dims,
+                        lengthscale_constraint=GreaterThan(1e-06),
+                    )
+                )
+            )
+
+            prod_kernel = (
+                ScaleKernel(
+                    cont_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(ord_dims),
+                        active_dims=ord_dims,
+                    )
+                )
+                * ScaleKernel(
+                    mol_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(mol_dims),
+                        active_dims=mol_dims,
+                    )
+                )
+                * ScaleKernel(
+                    CategoricalKernel(
+                        batch_shape=aug_batch_shape,
+                        ard_num_dims=len(cat_dims),  # type: ignore
+                        active_dims=cat_dims,
+                        lengthscale_constraint=GreaterThan(1e-06),
+                    )
+                )
+            )
+            covar_module = sum_kernel + prod_kernel
+
+        super().__init__(
+            train_X=train_X,
+            train_Y=train_Y,
+            likelihood=likelihood,
+            covar_module=covar_module,
+            outcome_transform=outcome_transform,
+            input_transform=input_transform,
+        )
+
+
+class MixedTanimotoGPSurrogate(BotorchSurrogate, TrainableSurrogate):
+    def __init__(
+        self,
+        data_model: DataModel,
+        **kwargs,
+    ):
+        self.continuous_kernel = data_model.continuous_kernel
+        self.categorical_kernel = data_model.categorical_kernel
+        self.molecular_kernel = data_model.molecular_kernel
+        self.scaler = data_model.scaler
+        super().__init__(data_model=data_model, **kwargs)
+
+    model: Optional[MixedTanimotoGP] = None
+    _output_filtering: OutputFilteringEnum = OutputFilteringEnum.ALL
+    training_specs: Dict = {}
+
+    def _fit(self, X: pd.DataFrame, Y: pd.DataFrame):
```

Comment:
This method has a relatively high complexity, I would propose to rip it apart in smaller helper functions like:
- `_get_categorical_features_and_indices`
- `_get_molecular_features_and_indices`
and so forth ...

One can then also write individual tests for the atomic methods.

Reply:
Thanks for the suggestion here. I have created functions that find continuous, categorical, or molecular features, and another general function that can identify the index of features. There can be found in the `data_models/domain/features.py` file in the Inputs class with function names: `get_molecular_features`, `get_continuous_features`, `get_categorical_features`, and `get_feature_indices`.

 Please note that the `get_molecular_features` and `get_categorical_features` functions are static methods but @simonsung06 felt that they would be best in the same "location" as  `get_continuous_features`. Happy for any input here from you :)